src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs

```diff
@@ -51,15 +42,137 @@ public class CryptoProviderFactory
         /// </summary>
         public static CryptoProviderFactory Default;
 
+        /// <summary>
+        /// Extensibility point for custom crypto support application wide.
+        /// </summary>
+        /// <remarks>By default, if set, <see cref="ICryptoProvider.IsSupportedAlgorithm(string, object[])"/> will be called before crypto operations.
+        /// If true is returned, then this will be called for operations.</remarks>
+        public ICryptoProvider CustomCryptoProvider { get; set; }
+
+        /// <summary>
+        /// Static constructor that initializes the default <see cref="CryptoProviderFactory"/>.
+        /// </summary>
         static CryptoProviderFactory()
         {
             Default = new CryptoProviderFactory();
         }
 
         /// <summary>
-        /// Delegate to resolve <see cref="AsymmetricAlgorithm"/> to use when signing or verifying token.
+        /// Default constructor for <see cref="CryptoProviderFactory"/>.
+        /// </summary>
+        public CryptoProviderFactory() { }
+
+        /// <summary>
+        /// Constructor that creates a deep copy of given <see cref="CryptoProviderFactory"/> object.
+        /// </summary>
+        /// <param name="other"><see cref="CryptoProviderFactory"/> to copy from.</param>
+        public CryptoProviderFactory(CryptoProviderFactory other)
+        {
+            if (other == null)
+                throw LogHelper.LogArgumentNullException(nameof(other));
+
+            CustomCryptoProvider = other.CustomCryptoProvider;
+        }
+
+        /// <summary>
+        /// Answers if an algorithm is supported
+        /// </summary>
+        /// <param name="algorithm">the name of the crypto algorithm</param>
+        /// <returns></returns>
+        public virtual bool IsSupportedAlgorithm(string algorithm)
+        {
+            if (CustomCryptoProvider != null && CustomCryptoProvider.IsSupportedAlgorithm(algorithm))
+                return true;
+
+            return IsSupportedHashAlgorithm(algorithm);
+        }
+
+        /// <summary>
+        /// Answers if an algorithm is supported
         /// </summary>
-        public AsymmetricAlgorithmResolver AsymmetricAlgorithmResolver { get; set; }
+        /// <param name="algorithm">the algorithm to use</param>
+        /// <param name="key">the <see cref="SecurityKey"/></param>
+        /// <returns></returns>
+        public virtual bool IsSupportedAlgorithm(string algorithm, SecurityKey key)
+        {
+            if (CustomCryptoProvider != null && CustomCryptoProvider.IsSupportedAlgorithm(algorithm, key))
+                return true;
+
+            if (key as RsaSecurityKey != null)
+                return IsRsaAlgorithmSupported(algorithm);
+
+            if (key as X509SecurityKey != null)
+                return IsRsaAlgorithmSupported(algorithm);
```

Comment:
I know DSA/ECDsa certificates are not supported ATM by IdentityModel, but this method will return `true` for any RSA-based algorithm even if the certificate is not a RSA certificate. Please fix that.